include/deal.II/base/mpi.h

```diff
@@ -268,6 +272,109 @@ namespace Utilities
       MPI_Comm comm;
     };
 
+    /**
+     * This class represents a mutex to guard a critical section for a set of
+     * processors in a parallel computation using MPI.
+     *
+     * The lock() commands waits until all MPI ranks in the communicator (given
+     * in the constructor of this class) have released a previous lock using
+     * unlock().
+     *
+     * A typical usage involves guarding a critical section using a lock guard:
+     * @code
+     * {
+     *   static CollectiveMutex mutex(comm);
+     *   std::lock_guard<CollectiveMutex> lock(mutex);
+     *   // [ critical code to be guarded]
+     * }
+     * @endcode
+     *
+     * Here, the critical code will finish on all processors before the mutex
+     * can be acquired again (for example by a second execution of the block
+     * above. The critical code block typically involves MPI communication that
+     * would yield incorrect results without the lock. For example, if the code
+     * contains nonblocking receives with MPI_ANY_SOURCE, packets can be
+     * confused between iterations.
+     *
+     * Note that the mutex needs to be the same instance between calls to the
+     * same critical region. While not required, this can be achieved by making
+     * the instance static (like in the example above). The variable can also be
+     * a global variable, or a member variable of the object to which the
+     * executing function belongs.
+     *
+     * The implementation uses an MPI barrier in the unlock() function. For MPI
+     * version 3 or newer, this barrier is executed in a non-blocking variant
+     * and we wait for completion only in the next call to lock() (which, in the
+     * example above, is called by the constructor of the `lock` object).
+     */
+    class CollectiveMutex
+    {
+    public:
+      /**
+       * This helper class provides a scoped lock for the CollectiveMutex.
+       */
+      class ScopedLock
+      {
+      public:
+        explicit ScopedLock(CollectiveMutex &mutex, const MPI_Comm &comm)
+          : mutex(mutex)
+          , comm(comm)
+        {
+          mutex.lock(comm);
+        }
+
+        ~ScopedLock()
+        {
+          mutex.unlock(comm);
+        }
+
+      private:
+        CollectiveMutex &mutex;
+        const MPI_Comm & comm;
+      };
+
+      /**
+       * Constructor of this class.
+       */
+      explicit CollectiveMutex();
+
+      /**
+       * Destroy the mutex. Assumes the lock is not currently held.
+       */
+      ~CollectiveMutex();
+
+      /**
+       * Aquire the mutex and wait until we can do so.
+       *
+       * This is a collective call that needs to be executed by all processors
+       * in the communicator.
```

Comment:
is this actually true (for the `lock()` function here)? I don't actually think so!